### ISSUE TITLE
lists.elem_at mutates its input

### ISSUE THREAD

**Issue by mboes** [tweag/nickel#534]
**Describe the bug**

The meat of the `elem_at` primitive is here:

https://github.com/tweag/nickel/blob/be46e932c90283f7b6699c54649f6bfff9d809d6/src/operation.rs#L1708-L1711

Where I would have expected a simple non-mutating indexing of the array, we instead use `swap_remove()`, which mutates the input to `elem_at`. Now, it's suprisingly hard to observe the effects of this, meaning that we are probably doing deep copies of closures somewhere. So while I can't yet observe incorrect semantics for `elem_at`, the implementation looks very fishy. We are either copying way too much or are getting the semantics wrong. Either way that's a bug. Furthermore, surely indexing should be how we implement `lists.elem_at`? That would make for clearer code. There is currently no explanation for why mutation is necessary.

**Additional context**
Here's an attempt to observe incorrect semantics:

```
nickel> let xs = [1,2,3]
nickel> lists.elem_at 0 xs
1
nickel> :print xs
[ 1, 2, 3 ]
```

Based on the semantics of `swap_remove()`, one would expect to see `[3, 2]` in the last line. The fact that we're not seeing that suggests to me that we are copying `xs` somewhere. We really shouldn't - `xs` could be a large array.

**Comment by yannham** [comment]
Currently all the pointers in the AST are owned (`Box`). So any sharing (like the evaluation of a non-linear occurrence of a  variable) incur a deep-copy, and prim op like `elem_at` get fully owned data. At this point, even if it may be slightly meaningless given all the previous deep copies, `swap_remove` avoids one additional deep copy.

This is intended to be solved by #426, but the first review haven't considered this point yet, so this is a good idea to keep this issue open until #426 is merged and we are sure we didn't keep any undue deep-copying due to the previous absence of sharing.

**Comment by yannham** [comment]
Closed in #426.
